### PR TITLE
added code to collect data from IMU sensor

### PR DESCRIPTION
**Pull request by avadesh02**
Added functionality to collect data from IMU. Also, is it possible to make the initialization optional so that the sensor are not initialised when they are not connected to the hardware?

### CODE REVIEW

**Comment by jviereck** [comment]
I wonder if the `Teststand` constructor should take arguments like `Teststand(single_can_card=[false]/true, with_ati=false/[true], with_imu=[false]/true)`?

File: demos/demo_teststand.cpp [hunk 1]
```diff
@@ -17,7 +17,9 @@ using namespace blmc_robots;
 
 static THREAD_FUNCTION_RETURN_TYPE control_loop(void* robot_void_ptr)
 {
-    Teststand& robot = *(static_cast<Teststand*>(robot_void_ptr));
```
Comment (MaximilienNaveau) [inline demos/demo_teststand.cpp]: When you create a real time thread it is strongly discourage to create the memory inside the thread. Do you have a reasonable reason for doing this? Otherwise please revert this change and create the robot in the main.

File: demos/single_leg.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 try:
     while True:
         leg.acquire_sensors()
-        des_pos = leg.get_slider_positions()
+        des_pos = leg.get_slider_positions() - 0.5
```
Comment (MaximilienNaveau) [inline demos/single_leg.py]: @jviereck is this reasonable?

File: include/blmc_robots/teststand.hpp [hunk 3]
```diff
@@ -12,6 +12,8 @@
 #include <blmc_robots/common_header.hpp>
 #include <blmc_robots/blmc_joint_module.hpp>
 #include <AtiFTSensor.h>
+#include <imu-core/imu_3DM_GX5_25.hpp>
+#include <stdio.h>
```
Comment (MaximilienNaveau) [inline include/blmc_robots/teststand.hpp]: Do not use the stdio. Use the real_time_tools/iostream.hpp which provides you with the rt_printf function.

File: src/teststand.cpp [hunk 4]
```diff
@@ -62,14 +62,14 @@ void Teststand::initialize()
   can_buses_[0] = std::make_shared<blmc_drivers::CanBus>("can0");
   can_motor_boards_[0] = std::make_shared<blmc_drivers::CanBusMotorBoard>(can_buses_[0]);
 
-  can_buses_[1] = std::make_shared<blmc_drivers::CanBus>("can1");
-  can_motor_boards_[1] = std::make_shared<blmc_drivers::CanBusMotorBoard>(can_buses_[1]);
+  // can_buses_[1] = std::make_shared<blmc_drivers::CanBus>("can1");
```
Comment (MaximilienNaveau) [inline src/teststand.cpp]: Let us try to find a reasonable way of handling the parametrization of the Testand hardware.

**Comment by MaximilienNaveau** [comment]
> I wonder if the `Teststand` constructor should take arguments like `Teststand(single_can_card=[false]/true, with_ati=false/[true], with_imu=[false]/true)`?

That might be a very good way to do it. And then upon creation you define this argument the way you want:
- Yaml paramters
- Executable arguments
- ...

File: demos/demo_teststand.cpp (hunk 1 above)
Comment (jviereck) [reply demos/demo_teststand.cpp]: > When you create a real time thread it is strongly discourage to create the memory inside the thread.

Can you explain why? I understand you shouldn't allocate on the main control loop but I don't see why doing it at the beginning of the rt thread before doing control is a problem.

File: demos/single_leg.py (hunk 2 above)
Comment (jviereck) [reply demos/single_leg.py]: Note sure.

@avadesh02 , why did you change this?

File: demos/demo_teststand.cpp (hunk 1 above)
Comment (jviereck) [reply demos/demo_teststand.cpp]: > Do you have a reasonable reason for doing this?

Similar like what I found when using the master board sdk: When initializing the object on the main thread and then start using the object passed via a pointer from the rt thread, things don't work. A similar behavior happend when we tried to read from the IMU.

We should investigate this further but I was not sure if we need to do it in this PR or not.

@MaximilienNaveau, @vincentberenz , can you look into why init from main thread and using from rt thread is causing problems? In our case we init the object and then spawn the rt thread. I wonder if spawning the rt thread takes some time. If the device expects to have constant communication after initialization, then this delay might cause the communication to break. This could explain what we observe.

Comment (MaximilienNaveau) [reply demos/demo_teststand.cpp]: I can check, but the imu code is simple, upon init I start a real time thread that reads the imu. When you do a get on the imu object you simply copy the last read results. so the communication is never interrupted. Except if creating a real time thread preempt everything else...

Comment (MaximilienNaveau) [reply demos/demo_teststand.cpp]: > > When you create a real time thread it is strongly discourage to create the memory inside the thread.
> 
> Can you explain why? I understand you shouldn't allocate on the main control loop but I don't see why doing it at the beginning of the rt thread before doing control is a problem.

This is what is recommended by the official rt_preempt repos in their development guide. They admit that creating the memory in the real time thread should not cause too much trouble but they advice against doing it. Ideally the real time thread just do some operation into some already initialized objects.
Last paragraph:
https://wiki.linuxfoundation.org/realtime/documentation/howto/applications/memory#memory_locking

**Comment by avadesh02** [comment]
Hi,

@max - The stdio is not being used anywhere. It can be removed. I imported
it while I was debugging something in the code.

@julian - I don't remember modifying the slider values. Maybe I did it
sometime for a demo. Should I remove it or can you do it?

Also for the IMU drivers is it possible to shift all the arguments expected
in the constructor into the initialize method? That way we don't have the
complication of creating a shared_ptr. Also it makes more sense to me that
creation of real-time thread etc ..happens when we call initialise and not
when we declare a variable of the class. Does this makes sense?

Regards,
Avadesh

**Comment by MaximilienNaveau** [comment]
Hi @avadesh02 ,

This is your PR so you should make the different fix.
For the change of API for the IMU, I have nothing against, just open a PR on imu-core.

In any case if you have trouble with executing the real-time thread it means there is something wrong in the code. I have been implementing tones of different real-time thread without any issues.
Maybe there is something wrong in imu-core or in your use of it.

For the slider I do think the default behavior should give you the [min=0, max=1] values of the raw sensors. The rest is application dependent.

Best
Max

File: demos/demo_teststand.cpp (hunk 1 above)
Comment (jviereck) [reply demos/demo_teststand.cpp]: Following recommendations is good. However, it doesn't work all the time. For instance, for tensorflow, I had to move the main initialization (and memory allocation) inside the rt_thread as well. When I tried to initialize the tensorflow object on the main thread and use the allocated objects from the rt_thread, things crashed.

See this commit: https://git-amd.tuebingen.mpg.de/amd-clmc/dg_tensorflow/commit/fc94de7fdc179d29f2a22975b198a7f458aaff9a

**Comment by jviereck** [comment]
Hi @avadesh02 ,

please use the GitHub ids (like @jviereck) instead of @julian. Otherwise you might reference other users then the ones in this conversation.

> @Julian - I don't remember modifying the slider values. Maybe I did it sometime for a demo. Should I remove it or can you do it?

Change it back to the way it's currently on `master`.

Best,
Julian
